crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -683,4 +689,30 @@ mod test {
         stage.run(&mut world);
         assert!(*world.get_resource::<bool>().unwrap(), "after test");
     }
+
+    #[test]
+    fn pr_2211() {
+        let mut world = World::default();
+
+        world.insert_resource(0u32);
+        world.insert_resource(State::new(MyState::S1));
+
+        let mut stage = SystemStage::parallel();
+
+        stage.add_system_set(State::<MyState>::get_driver());
+
+        stage.add_system_set(
+            SystemSet::on_in_stack_update(MyState::S1).with_system(
+                (|mut r: ResMut<u32>, mut s: ResMut<State<MyState>>| {
+                    *r += 1;
+                    s.push(MyState::S2).ok();
+                })
+                .system(),
+            ),
+        );
+        stage.run(&mut world);
```

Comment:
I don't really understand why this should trigger our system twice. Could you explain?

Reply:
As the first run create a state change, systems with a valid run criteria are rerun in the same frame.
Other runs that the first don't actually change the state, the methods return an `Err` which is ignored.

Reply:
@mockersf I understand that this is unrelated to the PR, but is this expected behavior? I faced the same too in my project.